File: utils/logger.js
import * as fs from 'fs';
import * as path from 'path';
import { getOutputPath } from '../src/path-resolver.js';

const LOG_DIR = getOutputPath();
const ERROR_LOG_PATH = path.join(LOG_DIR, 'errors.log');

export const LogLevel = {
    DEBUG: 0,
    INFO: 1,
    WARN: 2,
    ERROR: 3,
    QUIET: 4,
};

let currentLogLevel = LogLevel.INFO;

export function setLogLevel(level) {
    if (typeof level === 'string') {
        switch (level.toLowerCase()) {
            case 'debug':
                currentLogLevel = LogLevel.DEBUG;
                break;
            case 'info':
                currentLogLevel = LogLevel.INFO;
                break;
            case 'warn':
                currentLogLevel = LogLevel.WARN;
                break;
            case 'error':
                currentLogLevel = LogLevel.ERROR;
                break;
            case 'quiet':
            case 'silent':
                currentLogLevel = LogLevel.QUIET;
                break;
            default:
                currentLogLevel = LogLevel.INFO;
        }
    } else {
        currentLogLevel = level;
    }
}

export function getLogLevel() {
    return currentLogLevel;
}

export function shouldLog(level) {
    return level >= currentLogLevel;
}

function formatTimestamp() {
    return new Date().toISOString();
}

function ensureLogDir() {
    if (!fs.existsSync(LOG_DIR)) {
        fs.mkdirSync(LOG_DIR, { recursive: true });
    }
}

export function log(level, category, message, data = null) {
    if (!shouldLog(level)) return;

    const timestamp = formatTimestamp();
    const levelNames = ['DEBUG', 'INFO', 'WARN', 'ERROR'];
    const levelName = levelNames[level] || 'UNKNOWN';

    const formatted = `[${timestamp}] [${levelName}] [${category}] ${message}`;

    if (level >= LogLevel.ERROR) {
        console.error(formatted);
    } else {
        console.log(formatted);
    }

    if (data) {
        const dataStr = typeof data === 'object' ? JSON.stringify(data, null, 2) : String(data);
        console.log(dataStr);
    }

    if (level >= LogLevel.ERROR) {
        appendToErrorLog(formatted, data);
    }
}

export function debug(category, message, data = null) {
    log(LogLevel.DEBUG, category, message, data);
}

export function info(category, message, data = null) {
    log(LogLevel.INFO, category, message, data);
}

export function warn(category, message, data = null) {
    log(LogLevel.WARN, category, message, data);
}

export function error(category, message, data = null) {
    log(LogLevel.ERROR, category, message, data);
}

export function appendToErrorLog(message, data = null) {
    ensureLogDir();

    try {
        let logEntry = message;
        if (data) {
            logEntry += '\n  Data: ' + (typeof data === 'object' ? JSON.stringify(data) : String(data));
        }
        logEntry += '\n';

        fs.appendFileSync(ERROR_LOG_PATH, logEntry);
    } catch (err) {
        console.error('Failed to write to error log:', err.message);
    }
}

export function getErrorLog() {
    ensureLogDir();

    if (!fs.existsSync(ERROR_LOG_PATH)) {
        return [];
    }

    try {
        const content = fs.readFileSync(ERROR_LOG_PATH, 'utf-8');
        return content.split('\n').filter(line => line.trim());
    } catch {
        return [];
    }
}

export function clearErrorLog() {
    ensureLogDir();

    if (fs.existsSync(ERROR_LOG_PATH)) {
        fs.unlinkSync(ERROR_LOG_PATH);
    }
}

export function logPipelineStep(stepName, duration, success = true, error = null) {
    const entry = {
        timestamp: formatTimestamp(),
        step: stepName,
        duration_ms: duration,
        success,
        error: error?.message || null,
    };

    ensureLogDir();

    const logFile = path.join(LOG_DIR, 'pipeline_history.jsonl');
    try {
        fs.appendFileSync(logFile, JSON.stringify(entry) + '\n');
    } catch (err) {
        console.error('Failed to write pipeline history:', err.message);
    }

    return entry;
}

File: utils/logger.py
from __future__ import annotations

import json
import os
from datetime import UTC
from typing import Any

from src.path_resolver import get_output_path

LOG_DIR = get_output_path()
ERROR_LOG_PATH = os.path.join(LOG_DIR, "errors.log")


class LogLevel:
    DEBUG = 0
    INFO = 1
    WARN = 2
    ERROR = 3
    QUIET = 4


_current_log_level: int = LogLevel.INFO


def set_log_level(level: str | int) -> None:
    global _current_log_level
    if isinstance(level, str):
        mapping = {
            "debug": LogLevel.DEBUG,
            "info": LogLevel.INFO,
            "warn": LogLevel.WARN,
            "error": LogLevel.ERROR,
            "quiet": LogLevel.QUIET,
            "silent": LogLevel.QUIET,
        }
        _current_log_level = mapping.get(level.lower(), LogLevel.INFO)
    else:
        _current_log_level = level


def get_log_level() -> int:
    return _current_log_level


def should_log(level: int) -> bool:
    return level >= _current_log_level


def _format_timestamp() -> str:
    from datetime import datetime, timezone

    return datetime.now(UTC).isoformat()


def _ensure_log_dir() -> None:
    if not os.path.exists(LOG_DIR):
        os.makedirs(LOG_DIR, exist_ok=True)


def log(level: int, category: str, message: str, data: Any = None) -> None:
    if not should_log(level):
        return

    timestamp = _format_timestamp()
    level_names = ["DEBUG", "INFO", "WARN", "ERROR"]
    level_name = level_names[level] if level < len(level_names) else "UNKNOWN"

    formatted = f"[{timestamp}] [{level_name}] [{category}] {message}"

    if level >= LogLevel.ERROR:
        print(formatted, flush=True)
    else:
        print(formatted, flush=True)

    if data is not None:
        data_str = json.dumps(data, indent=2) if isinstance(data, dict) else str(data)
        print(data_str, flush=True)

    if level >= LogLevel.ERROR:
        _append_to_error_log(formatted, data)


def debug(category: str, message: str, data: Any = None) -> None:
    log(LogLevel.DEBUG, category, message, data)


def info(category: str, message: str, data: Any = None) -> None:
    log(LogLevel.INFO, category, message, data)


def warn(category: str, message: str, data: Any = None) -> None:
    log(LogLevel.WARN, category, message, data)


def error(category: str, message: str, data: Any = None) -> None:
    log(LogLevel.ERROR, category, message, data)


def _append_to_error_log(message: str, data: Any = None) -> None:
    _ensure_log_dir()
    try:
        log_entry = message
        if data is not None:
            log_entry += "\n  Data: " + (json.dumps(data) if isinstance(data, dict) else str(data))
        log_entry += "\n"
        with open(ERROR_LOG_PATH, "a", encoding="utf-8") as f:
            f.write(log_entry)
    except Exception as err:
        print(f"Failed to write to error log: {err}", flush=True)


def get_error_log() -> list[str]:
    _ensure_log_dir()
    if not os.path.exists(ERROR_LOG_PATH):
        return []
    try:
        with open(ERROR_LOG_PATH, encoding="utf-8") as f:
            content = f.read()
        return [line for line in content.split("\n") if line.strip()]
    except Exception:
        return []


def clear_error_log() -> None:
    _ensure_log_dir()
    if os.path.exists(ERROR_LOG_PATH):
        os.unlink(ERROR_LOG_PATH)


def log_pipeline_step(
    step_name: str,
    duration: float,
    success: bool = True,
    error: Exception | str | None = None,
) -> dict[str, Any]:
    entry = {
        "timestamp": _format_timestamp(),
        "step": step_name,
        "duration_ms": duration,
        "success": success,
        "error": str(error) if error else None,
    }

    _ensure_log_dir()
    log_file = os.path.join(LOG_DIR, "pipeline_history.jsonl")
    try:
        with open(log_file, "a", encoding="utf-8") as f:
            f.write(json.dumps(entry) + "\n")
    except Exception as err:
        print(f"Failed to write pipeline history: {err}", flush=True)

    return entry
